zookeeper-server/src/main/java/org/apache/zookeeper/server/RateLogger.java

```diff
@@ -65,14 +65,29 @@ public void rateLimitLog(String newMsg) {
     }
 
     /**
-     * In addition to the message, it also takes a value.
+     * Any new message is logged once to the underlying logger.
+     *
+     *
+     * If subsequent messages remain the same as the previous
+     * they are not logged unless a specified time interval has elapsed.
+     * A subsequent message that is different to the previous also logs the
+     * previous message.
+     * At any time the current message can be logged using {@link #flush()}.
+     * <p>
+     * Messages are written to log with format '
+     *
+     *
+     * @param newMsg the message to log;
+     * @param value the value provided while logging the message
+     *                     message value; Optional
      */
     public void rateLimitLog(String newMsg, String value) {
         long now = Time.currentElapsedTime();
         if (newMsg.equals(msg)) {
             ++count;
-            this.value = value;
+            this.value = value;  // should this go in an else block?
             if (now - timestamp >= LOG_INTERVAL) {
+                // log previous message and value
                 flush();
                 msg = newMsg;
```

Comment:
Why do you need this change? Time interval has elapsed, reset count?

Reply:
If the time interval has elapsed call flush writing the existing value and current message count then save the message with new value provided as the first message received after the flush with count 1.

This change makes it consistent with an explicit flush, which writes the current message along with exiting value and current message count.

Reply:
- Haha, not very persuasive for me:). 
- And for test, when time interval has elapsed and at the same time, `flush()` is called explicitly, the tests will become complicated.

```
e.g:
A Permutation and Combination of these three lines
currentTime.set(DEFAULT_LOGGING_INTERVAL + 1L);
rateLogger.rateLimitLog("test message one.", "value-four");
rateLogger.flush();
```

Reply:
Hey @maoling, the permutation you describe above is pretty much the 'recurringMessageTimedImplicitFlush' test

The following permutation prints the message with "value-four" immediately since flush clears everything and sets the held msg to 'null':
```
rateLogger.flush();
currentTime.set(DEFAULT_LOGGING_INTERVAL + 1L);
rateLogger.rateLimitLog("test message one.", "value-four");
```
Happy to revert if you're not comfortable.

Reply:
@jono-morris 
Take it easy, let's wait here for other people's review and comment

Reply:
@jono-morris 
Take it easy, let's wait here for other people's review and comment